benchmarks/distributed/test_distributed.py

```diff
@@ -0,0 +1,194 @@
+import ray
+import ray.autoscaler.sdk
+from ray.test_utils import Semaphore
+from ray.util.placement_group import placement_group, remove_placement_group
+
+from time import sleep, perf_counter
+from tqdm import tqdm, trange
+
+TEST_NUM_NODES = 64
+MAX_ACTORS_IN_CLUSTER = 10000
+MAX_RUNNING_TASKS_IN_CLUSTER = 10000
+MAX_PLACEMENT_GROUPS = 1000
+MAX_NUM_NODES = 250
+
+
+def scale_to(target):
+    while len(ray.nodes()) != target:
+        ray.autoscaler.sdk.request_resources(bundles=[{"node": 1}] * target)
+        print(f"Current # nodes: {len(ray.nodes())}, target: {target}")
+        print("Waiting ...")
+        sleep(5)
+
+
+def test_nodes():
+    scale_to(MAX_NUM_NODES)
+    assert len(ray.nodes()) == MAX_NUM_NODES
+    # Treat this as a trivial task to ensure the nodes are all functioning
+    test_max_running_tasks()
+
+
+def test_max_actors():
+    # TODO (Alex): Dynamically set this based on number of cores
+    cpus_per_actor = 0.25
+
+    @ray.remote(num_cpus=cpus_per_actor)
+    class Actor:
+        def foo(self):
+            pass
+
+    actors = [
+        Actor.remote() for _ in trange(MAX_ACTORS_IN_CLUSTER, desc="Launching actors")
+    ]
+
+    for actor in tqdm(actors, desc="Ensuring actors have started"):
+        assert ray.get(actor.foo.remote()) is None
+
+
+def test_max_running_tasks():
+    counter = Semaphore.remote(0)
+    blocker = Semaphore.remote(0)
+    total_resource = ray.cluster_resources()["node"]
+
+    # Use this custom node resource instead of cpus for scheduling to properly loadbalance. CPU borrowing could lead to unintuitive load balancing if we used num_cpus.
+    resource_per_task = total_resource / MAX_RUNNING_TASKS_IN_CLUSTER
+
+    @ray.remote(num_cpus=0, resources={"node": resource_per_task})
+    def task(counter, blocker):
+        sleep(10)
```

Comment:
This might not actually be running concurrent tasks right? Can you instead have the tasks block forever and use num_cpus=1? They can be killed at the end of the test.

Reply:
hmm it's kinda hard to cancel a running task right? Are you suggesting ending the job?

Reply:
btw we also don't get to confirm that the task actually executed if we do this (i originally tried to have the tasks all communicated via a semaphore actor, but a single actor couldn't communicate with 10k workers at once)

Reply:
The problem is this definitely isn't testing the right thing. The tasks might all just run on the head node or something like that.

How about we at least check the number of concurrent tasks by polling ray.available_resources() (checking to see how many CPUs max over the run)?